### DISCUSSION TITLE
How to set corner radius dependent on height

### DISCUSSION THREAD

**Discussion by llfab** [AvaloniaUI/Avalonia#10566] [Q&A] [answered]
I created my own button derived user control and want to create a style that has a round left and right edge, hence, CornerRadius = 1/2 * height or so.

```csharp
  <!-- Major Style -->
  <Style Selector="ucp|DefaultButton.majorButton">
    
    <Style Selector="^ /template/ Border#border">
      <Setter Property="BorderThickness" Value="3" />
      <Setter Property="CornerRadius" Value="100" /> <!-- Should be dependent on height -->
      
    </Style>
  </Style>
```
It shall look like this:

![image](https://user-images.githubusercontent.com/31031996/223114378-0b48eaee-d41f-4c1f-9979-a430d65f9d84.png)

**Comment by llfab** [comment, marked as answer]
I tried ```{Binding $self.Bounds.Height}``` but I guess there is a type mismatch. Maybe a Converter would help, but is there not easier way???

**Reply by llfab** [reply]
This is what has worked but is pretty long-ish:
```csharp

public class DoubleCornerRadiusConverter : IValueConverter
{
    private static readonly Trace s_trace = Trace.ForType<DoubleCornerRadiusConverter>();

    public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
    {
        if (value is not double doubleValue)
        {
            return value;
        }

        if (targetType != typeof(CornerRadius))
        {
            return value;
        }

        return new CornerRadius(doubleValue);
    }
}

...

<Styles.Resources>
  <con:DoubleCornerRadiusConverter x:Key="cornerRadiusConverter" />
</Styles.Resources>

...

<Setter Property="CornerRadius" Value="{Binding $self.Bounds.Height, Converter={StaticResource cornerRadiusConverter}}" /> 
```

**Reply by timunie** [reply]
You can place it in a control lib if you want to, then it's re-usable anywhere. Also remember to divide the height by 2 for the correct result. 

See for referece: https://github.com/PieroCastillo/Aura.UI/blob/master/src/Aura.UI/Converters/BoundsToCornerRadiusConverter.cs 

@PieroCastillo I hope you don't mind that I share your code here.

**Reply by llfab** [reply]
Thanks. Here is an alternative which lets you define the ratio from ConverterParameter in XAML

```csharp
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
        {
            if (value is not Rect rectValue)
            {
                return new CornerRadius(0);
            }

            if (targetType != typeof(CornerRadius))
            {
                return new CornerRadius(0);
            }

            double relativeShortEdgeValue = 0.5;
            if (parameter != null)
            {
                relativeShortEdgeValue = System.Convert.ToDouble(parameter);
            }

            double shortEdge = Math.Min(rectValue.Width, rectValue.Height);
            double cornerRadius = shortEdge * relativeShortEdgeValue;

            return new CornerRadius(cornerRadius);
        }
```

**Comment by rabbitism** [comment]
Setting corner radius to a very large number looks exactly the same as you want.

**Reply by llfab** [reply]
I did notice that as well. However, I found that not right to set it to e.g. 10,000 or so...

**Reply by lukaszossowski** [reply]
Since it gets clipped anyway... why not? And I'm guessing a much smaller number would suffice, not that it should matter. It might actually be faster this way than using converter.

**Reply by llfab** [reply]
I wonder whether that behavior is by accident or by design. Could we not also want something like this with a larger corner radius?
![grafik](https://user-images.githubusercontent.com/31031996/223234427-ab211991-b2e7-440f-836f-73741f9009bd.png)

**Reply by timunie** [reply]
Yes, I agree. It's probably a bug and therefore the behavior may change in the future. So a converter is "saver" to use. Also the overhead of such a converter is negligible
